tools/java/runfiles/Runfiles.java

```diff
@@ -101,34 +177,105 @@ public static Runfiles create() throws IOException {
     } else {
-      return new DirectoryBased(getRunfilesDir(env));
+      return new DirectoryBased(getRunfilesDir(env), sourceRepository);
     }
   }
 
+  /**
+   * Returns a new instance that uses the provided source repository as a default for all calls to
+   * {@link #rlocation(String)}.
+   *
+   * <p>This is useful when receiving a {@link Runfiles} instance from a different Bazel
+   * repository. In this case, while the runfiles manifest or directory encoded in the instance
+   * should be used for runfiles lookups, the repository from which apparent repository names should
+   * be resolved needs to change.
+   *
+   * @param sourceRepository the canonical name of the Bazel repository relative to which apparent
+   *                         repository names should be resolved
+   * @return a new {@link Runfiles} instance identical to this one, except that calls to
+   * {@link #rlocation(String)} use the provided source repository.
+   */
+  public abstract Runfiles withSourceRepository(String sourceRepository);
+
   /**
    * Returns the runtime path of a runfile (a Bazel-built binary's/test's data-dependency).
    *
-   * <p>The returned path may not be valid. The caller should check the path's validity and that the
-   * path exists.
+   * <p>The returned path may not be valid. The caller should check the path's validity and that
+   * the path exists.
    *
    * <p>The function may return null. In that case the caller can be sure that the rule does not
    * know about this data-dependency.
    *
    * @param path runfiles-root-relative path of the runfile
    * @throws IllegalArgumentException if {@code path} fails validation, for example if it's null or
-   *     empty, or not normalized (contains "./", "../", or "//")
+   *                                  empty, or not normalized (contains "./", "../", or "//")
    */
   public final String rlocation(String path) {
+    return rlocation(path, this.sourceCanonical);
+  }
+
+  /**
+   * Returns the runtime path of a runfile (a Bazel-built binary's/test's data-dependency).
+   *
+   * <p>The returned path may not be valid. The caller should check the path's validity and that
+   * the path exists.
+   *
+   * <p>The function may return null. In that case the caller can be sure that the rule does not
+   * know about this data-dependency.
+   *
+   * @param path             runfiles-root-relative path of the runfile
+   * @param sourceRepository the canonical name of the Bazel repository relative to which apparent
+   *                         repository names in runfiles paths should be resolved. Overrides the
+   *                         repository set when creating this {@link Runfiles} instance.
+   * @throws IllegalArgumentException if {@code path} fails validation, for example if it's null or
+   *                                  empty, or not normalized (contains "./", "../", or "//")
+   */
+  public final String rlocation(String path, String sourceRepository) {
```

Comment:
What is the use case for rlocation with sourceRepostory function? If there isn't one please remove.

Canonical names seem to be hard enough to obtain, and users should not care about anything else than their current repo.

There's functionality missing to obtain other repo names. Let's not introduce it.

Reply:
In larger applications, `Runfiles` instances are typically created at the `lang_binary` level and then passed down via some kind of dependency injection pattern. This is crucial since finding the runfiles directory or manifest may require custom logic.

Then, all downstream consumers that need to access runfiles just work with an existing Runfiles instance. However, these downstream consumers may very well live in different repositories and would thus need to specify a custom `sourceRepository` to find *their* runfiles. This also came up in https://github.com/bazelbuild/proposals/pull/269#discussion_r927728158.

Reply:
That's a valid use case. 

It's quite error prone though to be passing Runfiles with `sourceRepository` set as you might cross repo boundary unaware. It would make sense to design and structure classes in a way to prevent this.

Reply:
If we started out with a clean slate, I would probably go for an API where `Create` doesn't take `sourceRepository` and returns a `Runfiles` that doesn't have any `rlocation` functions. Instead, users would need to call `withSourceRepository` on it to get an object of a *distinct* `RunfilesWithRepoContext` type that does have an `rlocation` method. Users would then be encouraged to only ever pass `Runfiles` into other repos, with no repo context. I don't see how to implement this in a backwards compatible way though - if you have any ideas, please let me know!

But it's ultimately impossible to prevent misuse even with these hypothetical changes. After all, if we could programatically determine repo boundaries, we wouldn't have the user pass in `sourceRepository` manually but just do it for them (which we can in some languages, e.g. Python, Bash and Go).

What I am thinking about as a potential followup is a debug mode, similar to what the Bash runfiles library offers: If a certain environment variable is set, runfiles lookups would fail with a descriptive exception if a repo name isn't found in the repo mapping but also doesn't look like a canonical repository name (which, with Bzlmod, are always either `_main` or contain a tilde, but of course that's an implementation detail we could only depend on from a runfiles library shipped with Bazel).

What do you think? Do you want me to make further changes here?

Reply:
You could add `static createPreloaded()` that returns a `PreloadedRunfiles` without `rlocation` and with `Runfiles withSourceRepository(String)` methods. Usability is not much different, i.e. `Runfiles runfiles = Runfiles.createPreloaded().withSourceRepository(AutoBazelRepository_MyClass.BAZEL_REPOSITORY);`

Reply:
@comius I pushed a new commit that implements this pattern. It is not complete yet (I haven't updated the docs).

If there are any changes to the user-facing API you would like me to make, could you let me know today? I have to finalize my BazelCon talk before Monday and would like to present this API in it.

Reply:
I also added `.unmapped()` in addition to `.withSourceRepository(String)` for the case where someone uses `rlocationpath` exclusively. Having them pass in a non-sense repo name would make the API confusing.

Reply:
> If there are any changes to the user-facing API you would like me to make, could you let me know today?

No, nothing out of the lines we're already talking about.

Reply:
@comius What do you think of renaming `createPreloaded` to just `preload`? That may differentiate this method more clearly from `create`, which has very different semantics.

Reply:
SGTM

Reply:
Stardoc shows some need to provide a public method that returns `RepositoryContext`. That would allow it to store labels in canonical form, using `Label.parseWithRepoContext` method.

Slightly related is, that it seems it's also possible to implement a method similar `rlocation` that takes an absolute label (but with non-canonical repos) as an input.

Reply:
`RepositoryMapping` and `Label` might be a dependency you don't want to bring into `Runfiles`. What also works for Stardoc is a more minimalistic `getCannonicalName(targetRepoApparentName)` or similar. This way you can hop around different repos by calling  `withSourceRepository(getCannonicalName(repoInLoad))`.

BTW one question. I was playing around with stardoc and I didn't get repo_mappings file in the exec root / sandbox. Is it possible it doesn't work yet or am I missing some flags?

Reply:
> `RepositoryMapping` and `Label` might be a dependency you don't want to bring into `Runfiles`. What also works for Stardoc is a more minimalistic `getCannonicalName(targetRepoApparentName)` or similar. This way you can hop around different repos by calling `withSourceRepository(getCannonicalName(repoInLoad))`.

I have thought about introducing such a function, but I don't know what it would be useful for other than implementing Stardoc. But that may be good enough of a reason to add it - after all, it would just have to be extracted out of this PR.

> BTW one question. I was playing around with stardoc and I didn't get repo_mappings file in the exec root / sandbox. Is it possible it doesn't work yet or am I missing some flags?

It's not staged properly yet, which I intend to fix with https://github.com/bazelbuild/bazel/pull/16652. When it's working, I will also add a small integration test to this PR.